
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  
  <!-- Licensed under the Apache 2.0 License -->
  <link rel="stylesheet" type="text/css" href="_static/fonts/open-sans/stylesheet.css" />
  <!-- Licensed under the SIL Open Font License -->
  <link rel="stylesheet" type="text/css" href="_static/fonts/source-serif-pro/source-serif-pro.css" />
  <link rel="stylesheet" type="text/css" href="_static/css/bootstrap.min.css" />
  <link rel="stylesheet" type="text/css" href="_static/css/bootstrap-theme.min.css" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
    <title>DisCVR Built-in Databases &#8212; DisCVR 0.1 documentation</title>
    <link rel="stylesheet" href="_static/guzzle.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Output Validation" href="validation.html" />
    <link rel="prev" title="Quick GUI overview" href="gui.html" />
  
   

  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="validation.html" title="Output Validation"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="gui.html" title="Quick GUI overview"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">DisCVR 0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="container-wrapper">

      <div id="mobile-toggle">
        <a href="#"><span class="glyphicon glyphicon-align-justify" aria-hidden="true"></span></a>
      </div>
  <div id="left-column">
    <div class="sphinxsidebar">
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <h2>Contents</h2>
    <div class="sidebar-localtoc">
      <ul>
<li><a class="reference internal" href="#">DisCVR Built-in Databases</a></li>
<li><a class="reference internal" href="#discvr-classification">DisCVR Classification</a></li>
<li><a class="reference internal" href="#classification-output">Classification Output</a></li>
</ul>

    </div>
  </div>
</div>
  <h4>Previous topic</h4>
  <p class="topless"><a href="gui.html"
                        title="previous chapter">Quick GUI overview</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="validation.html"
                        title="next chapter">Output Validation</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/classification.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div class="sidebar-block">
  <div class="sidebar-wrapper">
    <div id="main-search">
      <form class="form-inline" action="search.html" method="GET" role="form">
        <div class="input-group">
          <input name="q" type="text" class="form-control" placeholder="Search...">
        </div>
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div>
      
    </div>
  </div>
        <div id="right-column">
          
          <div role="navigation" aria-label="breadcrumbs navigation">
            <ol class="breadcrumb">
              <li><a href="index.html">Docs</a></li>
              
              <li>DisCVR Built-in Databases</li>
            </ol>
          </div>
          
          <div class="document clearer body">
            
  <div class="section" id="discvr-built-in-databases">
<h1>DisCVR Built-in Databases<a class="headerlink" href="#discvr-built-in-databases" title="Permalink to this headline">¶</a></h1>
<p>DisCVR tool has three virus <em>k</em>-mers (k=22) databases which are built into the DisCVR.jar:</p>
<ol class="arabic simple">
<li>Human Hemorrhagic virus dataset (hemorrhagic dataset)</li>
<li>Human Respiratory virus dataset (respiratory dataset)</li>
<li>Human Pathogenic viruses dataset (pathogenic dataset)</li>
</ol>
<p>The hemorrhagic and respiratory datasets consist of a list of selected circulating viruses.
The pathogenic datatset consist of a list of all human pathogenic viruses identified as biological agents by the
Health and Safety Executive (HSE) in the UK available at: <a class="reference external" href="http://www.hse.gov.uk/pubns/misc208.pdf">http://www.hse.gov.uk/pubns/misc208.pdf</a></p>
<p>Both the hemorrhagic and respiratory datasets overlap with the HSE database but are smaller in size and therefore are faster to use in classification.
Moreover, each dataset has a list of viral reference genomes that is associated with the viruses in the database. The reference genomes are used in the validation stage of DisCVR.
Information about the viruses in the databases can be obtained from the Database icon on the menu bar from the DisCVR GUI. The information includes the virus rank on the taxonomy tree,
the number of sequences used in the database build to represent it, and the accession number of its reference genome. Segmented viruses have their reference genome accession number listed
in decreasing order of the segment size. The accession numbers of the reference genomes are linked to their webpages on the NCBI, which can be viewed by clicking on the reference genome accession.</p>
</div>
<div class="section" id="discvr-classification">
<h1>DisCVR Classification<a class="headerlink" href="#discvr-classification" title="Permalink to this headline">¶</a></h1>
<p>The DisCVR GUI can be used to carry out a single sample classification. To launch the DisCVR GUI, either double click on DisCVR.jar, or
open a command prompt and type the following commands:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> full/path/to/DisCVR folder
java –jar full/path/to/DisCVR.jar
</pre></div>
</div>
<p>Once DisCVR is launched, there are 4 panels in the graphical interface:</p>
<ol class="arabic simple">
<li>Progress Bar: shows the progression of the classification process.</li>
<li>Input Sources: allows the user to choose the sample file and a <em>k</em>-mers database for classification.</li>
<li>Classification Results: shows the classification output in three sub-panels.</li>
<li>Progress Information: updates the user about classification output.</li>
</ol>
<p>To start a classification process, the user needs to first select the sample file they wish to
investigate. Currently, DisCVR supports .fasta, .fa, .fastq, and .fq formats. It allows for the selection
of one file hence, for paired-end files, it is recommended to concatenate both files. The second step is to
choose the database for the classification. The default setting is to select one of the three built-in databases.
This sets the <em>k</em>-mer size to 22 and removes all sample <em>k</em>-mers with entropy ≤ 2.5. On the other hand, if the user
chooses to use a customised database for the classification, then the box next to the <a class="reference internal" href="validation.html#customised-database"><span class="std std-ref">Customised Database</span></a> label
must be checked. The fields for this selection are then activated for the user to upload the customised database
file, which MUST be in the customisedDB folder that is in the same path as DisCVR.jar. The <em>k</em>-mer size for the
classification is extracted from the database name and the threshold to filter out low-entropy <em>k</em>-mers from the
sample should be entered in its corresponding field. The entropy threshold should be the same value used in the
build of the customised database which is recommended to be in the range of [0, 3.0]. If the users choose not to
specify a value for the entropy threshold then the default value of 2.5 is displayed in the field and used in the
classification.</p>
<p>Once, the database library is selected, the user clicks the classify button to start the classification process.</p>
<p>There are 4 stages in the classification process and the progress bar is updated after the completion of each stage.
Messages in relation to the output of each stage are displayed in the progress information panel.
This includes the number of reads in the sample, the number of distinct <em>k</em>-mers and their total counts in
the database and the sample file, the number of sample <em>k</em>-mers after removing single copies and low entropy
<em>k</em>-mers, and the number of classified <em>k</em>-mers and their total counts. The final message shows the number of
viruses with classified <a href="#id1"><span class="problematic" id="id2">*</span></a>k*mers found in the sample and the time taken to finish the classification process.
Figure 1 shows screenshots of the sample classification process using the DisCVR GUI.</p>
<div class="figure align-center" id="id3">
<a class="reference internal image-reference" href="_images/img1.png"><img alt="_images/img1.png" src="_images/img1.png" style="width: 800px;" /></a>
<p class="caption"><span class="caption-text">Figure2a: Screenshots of a sample classification process using DisCVR GUI. First the sample file is uploaded</span></p>
</div>
<div class="figure align-center" id="id4">
<a class="reference internal image-reference" href="_images/img2.png"><img alt="_images/img2.png" src="_images/img2.png" style="width: 800px;" /></a>
<p class="caption"><span class="caption-text">Figure2b: Screenshots of a sample classification process using DisCVR GUI. The <em>k</em>-mers database is then selected from the DisCVR database library</span></p>
</div>
<div class="figure align-center" id="id5">
<a class="reference internal image-reference" href="_images/img3.png"><img alt="_images/img3.png" src="_images/img3.png" style="width: 800px;" /></a>
<p class="caption"><span class="caption-text">Figure2b: Screenshots of a sample classification process using DisCVR GUI. The <em>k</em>-mers database is then selected from the user customised database*</span></p>
</div>
<div class="figure align-center" id="id6">
<a class="reference internal image-reference" href="_images/img4.png"><img alt="_images/img4.png" src="_images/img4.png" style="width: 800px;" /></a>
<p class="caption"><span class="caption-text">Figure2b: Screenshots of a sample classification process using DisCVR GUI. The classification results are displayed</span></p>
</div>
</div>
<div class="section" id="classification-output">
<h1>Classification Output<a class="headerlink" href="#classification-output" title="Permalink to this headline">¶</a></h1>
<p>At the end of the classification process, the progress information panel states the number of viruses with matched <em>k</em>-mers to the database.
Detailed information about the classification results are displayed on the centre panel.</p>
<p>The classification results panel consists of three sub-panels: Scoring, Summary, and Full Analysis. Once the classification is completed, a bar
chart showing up the top three viruses is displayed on the scoring panel. These are the viruses with the most number of classified distinct <em>k</em>-mers
found in the sample. The chart shows for each virus, the number of distinct <em>k</em>-mers that are specific to the virus as well as the number of shared
<em>k</em>-mers with other viruses in the sample, which are referred to as non-specific <em>k</em>-mers. The Summary panel gives a list of the viruses found in the
sample along with their number of specific and non-specific distinct <em>k</em>-mers. The Full Analysis panel shows a table with taxonomic and detailed
information about the viruses with classified <em>k</em>-mers from the sample. The results are shown in the table such that the first row shows the virus
with the highest number of <em>k</em>-mers found in the sample. However, the users can click on any column heading to sort out the results in the table
according to the information in the column. The full analysis table consists of 8 columns:</p>
<ol class="arabic simple">
<li>Virus Name: the scientific name for the virus taking from the NCBI names.dmp file.</li>
<li>Taxa ID: the taxonomy identification for the virus in the NCBI taxonomy tree.</li>
<li>Virus Rank: the rank of the virus according to the NCBI nodes.dmp file.</li>
</ol>
<p>4. Total counts of <em>k</em>-mers in DB: the total counts of <em>k</em>-mers that represent the virus in DisCVR’s
database.</p>
<p>5. No. of distinct Classified <em>k</em>-mers: the number of distinct <em>k</em>-mers that represent the virus in the
sample after removing single copies and low entropy <em>k</em>-mers and matched with the <em>k</em>-mers
database.</p>
<p>6. (%) of distinct Classified <em>k</em>-mers: the percentage of distinct classified <em>k</em>-mers that represent the
virus in the sample.</p>
<p>7. Total counts of Classified <em>k</em>-mers: the total number of <em>k</em>-mers that represent the virus in the
sample; some distinct <em>k</em>-mers can occur in the reads multiple times.</p>
<p>8. (%) of total classified <em>k</em>-mers: the percentage of the total number of <em>k</em>-mers that represent the
virus in the sample’s total number of <em>k</em>-mers.</p>
<p>The table can be saved as .csv file from the File icon on the GUI tool bar. Figure 2 shows screenshots
of the classification results sub-panels.</p>
<div class="figure align-center" id="id7">
<a class="reference internal image-reference" href="_images/img6.png"><img alt="_images/img6.png" src="_images/img6.png" style="width: 800px;" /></a>
<p class="caption"><span class="caption-text">Figure3a: Screenshots showing sample classification output on the results sub-panels. The scoring panel shows the three viruses with the highest number of matched <em>k</em>-mers</span></p>
</div>
<div class="figure align-center" id="id8">
<a class="reference internal image-reference" href="_images/img5.png"><img alt="_images/img5.png" src="_images/img5.png" style="width: 800px;" /></a>
<p class="caption"><span class="caption-text">Figure3b: Screenshots showing sample classification output on the results sub-panels. The summary panel</span></p>
</div>
<div class="figure align-center" id="id9">
<a class="reference internal image-reference" href="_images/img7.png"><img alt="_images/img7.png" src="_images/img7.png" style="width: 800px;" /></a>
<p class="caption"><span class="caption-text">Figure3c: Screenshots showing sample classification output on the results sub-panels. The full analysis panel with detailed information about the viruses found in the sample</span></p>
</div>
</div>


          </div>
            
  <div class="footer-relations">
    
      <div class="pull-left">
        <a class="btn btn-default" href="gui.html" title="previous chapter (use the left arrow)">Quick GUI overview</a>
      </div>
    
      <div class="pull-right">
        <a class="btn btn-default" href="validation.html" title="next chapter (use the right arrow)">Output Validation</a>
      </div>
    </div>
    <div class="clearer"></div>
  
        </div>
        <div class="clearfix"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="validation.html" title="Output Validation"
             >next</a> |</li>
        <li class="right" >
          <a href="gui.html" title="Quick GUI overview"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">DisCVR 0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
<script type="text/javascript">
  $("#mobile-toggle a").click(function () {
    $("#left-column").toggle();
  });
</script>
<script type="text/javascript" src="_static/js/bootstrap.js"></script>
  <div class="footer">
    &copy; Copyright 2019, Maha Maabar and Joseph Hughes. Created using <a href="http://sphinx.pocoo.org/">Sphinx</a>.
  </div>
  </body>
</html>